corehq/apps/accounting/models.py

```diff
@@ -2627,42 +2627,38 @@ def adjust_credit_balance(self, amount, is_new=False, note=None,
 
     @classmethod
     def get_credits_for_line_item(cls, line_item):
-        product_type = (
-            line_item.product_rate.product.product_type
-            if line_item.product_rate is not None else None
-        )
+        is_product = line_item.product_rate is not None
         feature_type = (
             line_item.feature_rate.feature.feature_type
             if line_item.feature_rate is not None else None
         )
 
-        for credit_line in cls.get_credits_by_subscription_and_features(
-            line_item.invoice.subscription,
-            product_type=product_type,
-            feature_type=feature_type,
-        ):
-            yield credit_line
-
-        if product_type is not None:
-            for credit_line in cls.get_credits_by_subscription_and_features(
-                line_item.invoice.subscription,
-                product_type=SoftwareProductType.ANY,
-            ):
-                yield credit_line
-
-        for credit_line in cls.get_credits_for_account(
-            line_item.invoice.subscription.account,
-            product_type=product_type,
-            feature_type=feature_type,
-        ):
-            yield credit_line
+        assert is_product or feature_type
```

Comment:
this is more about protecting it against misuse in the future than asserting about it's current behavior (which I am pretty confident is correct) so I'm ok with the hard assert.